Describe moving kotlinx.datetime.Instant to stdlib (#387)
Kotlin/KEEP: proposals/stdlib/instant.md

into the Kotlin Standard Library.

Additionally, it defines some interactions with `kotlin.time.Duration`, an
existing API for representing some number of seconds.

The API described here is available in `kotlinx-datetime` 0.6.0 in the form of
the `Instant` class and the `Clock` interface.
The goal of this proposal is to move this widely applicable functionality into
the Standard Library.

Overview
--------

### `Instant`

`Instant` denotes a moment in time, regardless of the observer.
Examples:

* The moment clocks showed 2024-08-05, 14:34:31 in New York.
* The moment Kotlin 2.0 was made available on Maven Central.
* The moment corresponding to the Unix timestamp `1722863508.231529`.
* The moment corresponding to the ISO 8601 string `2024-08-05T14:34:31+02:00`.

Non-examples:

* "2024-08-05, 14:34:31" is not an `Instant`,
  because people in Tokyo and Marocco would not observe this on their clocks
  simultaneously.
* "2024-08-05 UTC" is not an `Instant`, as it's too coarse.
  We define `Instant` values to have a precision of several seconds.

### `Clock`

A `Clock` implementation is a way to obtain the current `Instant`.

> The `Clock.SYSTEM` default implementation, available on all platforms, queries

Reviewer: Clock.SYSTEM -> Clock.System

**Example**: "the user was last online at 15:32" is calculated like this.

For the backend, `Instant` is enough to support this use case.
For the frontend, additionally, time zone information is needed.

#### Logging events

For logging, typically, the `Instant` is included and not the local datetime.
This makes it possible to merge logs across several computers, and if their
clocks are not too out of sync, a consistent view of event ordering can be
restored, regardless of which time zones these computers were configured to use.

**Example**: log entries in Linux store the number of milliseconds since the
start of the Unix epoch, and these can be represented as `Instant` values:

```sh
$ journalctl -o json | jq .__REALTIME_TIMESTAMP | head -n1
"1715603477525420"
```

**Example**: filesystems typically store information about when files were
created, modified, and/or accessed.
This information is stored as `Instant` values.

#### Determining time intervals between events

**Example**: "This user was last online three minutes ago".

**Example**: listing all files that were modified in the last 24 hours.

These calculations require knowing the current moment in time (accessed via

> `Clock.SYSTEM`) and the moment in time when some event happened.

Reviewer: Clock.SYSTEM -> Clock.System

// Arithmetics
println(runCatching {
    java.time.Instant.EPOCH
        .plusSeconds(Long.MAX_VALUE)
}) // Failure(java.time.DateTimeException: Instant exceeds minimum or maximum instant)
println(runCatching {
    kotlinx.datetime.Instant.fromEpochMilliseconds(0)
        .plus(kotlin.time.Duration.INFINITE)
}) // Success(+1000000000-12-31T23:59:59.999999999Z)
```

Where Java consistently enforces the Instant boundaries, we instead choose to
clamp the results to `Instant.MAX` or `Instant.MIN`, depending on the sign.

The idea is that when someone has years bigger than `1000000`, then surely
these values do not actually represent an `Instant` value in the business logic.
Short of a sci-fi fiction writer calculating the number of seconds to give as
input to a time machine to witness the extinction of dinosaurs
66 million years ago first-hand, there are barely any cases where such large
values are meaningful and should be correctly preserved.

Instead, we expect that such large values are simply sentinel `MAX`/`MIN` values
in various systems. Consider this code:

```kotlin
val maxTimestamp = Instant.fromEpochSeconds(MAX_POSTGRES_INTERVAL)

> val nextEvent =
>     timestamps.map(Instant::fromEpochSeconds)
>         filter { it > currentTime }.minOrNull() ?: maxTimestamp
> 
> if (nextEvent != maxTimestamp) {
>     println("The next event is $earliestEvent")

Reviewer 2: ```suggestion
val nextEvent =
    timestamps.map(Instant::fromEpochSeconds)
        .filter { it > currentTime }.minOrNull() ?: maxTimestamp

if (nextEvent != maxTimestamp) {
    println("The next event is $nextEvent")
```

The reason for this is our choice of the UTC-SLS time scale: in that time scale,
`Instant` values with `23:59:60Z` do not exist, and `23:59:59Z` never gets
omitted, and instead, the extra and missing seconds are emulated by speeding
the clock up and down.

Java supports parsing `23:59:60Z`, but when the offset is different from `Z`,
it does so incorrectly as of writing, allowing inserting leap seconds at
`23:59:60` regardless of the UTC offset and forbidding the second value `:60`
in all other combinations.

This may be an issue for compatibility: if another system produces
`23:59:60Z`, we will not be able to parse that.
In that case, the user can employ the parsing and formatting API in
`kotlinx-datetime` to decide how they want to handle leap seconds.
It's also worth noting that we didn't find a significant demand for leap
second handling in other parsing and formatting APIs, so this concern is mostly
theoretical.

#### The deprecated `now`

A common way to obtain the current `Instant` on the JVM is to call
`Instant.now()`:

* <https://grep.app/search?q=Instant.now%28%29> is 13000 hits as of writing.
* <https://grep.app/search?q=Clock.systemUTC%28%29.instant%28%29> is less than

>   100 hits.

Reviewer 3: Not sure this comparison is useful here since these expressions are equivalent. The point of the `Clock` abstraction is instead to facilitate inversion of control as noted in its section below. You'd have to compare it to those calling `Clock::instant` which is much harder to search for. If we assume people name their injected `Clock` as "clock", you get 1200 hits: https://grep.app/search?q=clock.instant%28%29. That's not exhaustive, but it's probably a majority.

It is not obvious to people who want to obtain the current `Instant` that a
class like `Clock` even exists, so we direct them to it using a deprecation with

> a proposed replacement.

Reviewer 2: Isn't documentation for the `Instant` class enough?

Author: If anyone does read it, it should be enough. `Instant.now()` is for establishing a smooth enough migration using the compiler errors that it's straightforward to arrive at correct code without reading anything, just with autocomplete.

We do not have any hard constraints on the implementation.
It is possible that `kotlinx-datetime` releases with its own implementation of
`java.io.Serializable` for `kotlinx.datetime.Instant`, but these formats don't
have to be compatible: whenever someone serializes `kotlinx.datetime.Instant`,
they expect to receive just that on deserialization.
When the code serializing `kotlinx.datetime.Instant` is migrated to
`kotlin.time.Instant`, the old serialization implementation is no longer
relevant.

### Additional platform-specific converters

On all platforms, there are entities fulfilling the same purpose as
`Instant`, and for some of them, we provide automatic conversion functions:

```kotlin
// JVM

public fun kotlin.time.Instant.toJavaInstant(): java.time.Instant

public fun java.time.Instant.toKotlinInstant(): kotlin.time.Instant
```

```kotlin
// JS

/**
 * Converts the [Instant] to an instance of JS [Date].
 *
 * The conversion is lossy: JS uses millisecond precision to represent dates,
 * and [Instant] allows for nanosecond resolution.
 */
public fun Instant.toJSDate(): Date

public fun Date.toKotlinInstant(): Instant
```

```kotlin

> // Darwin

Reviewer 3: According to the UUID KEEP, the stdlib is not specialized for various native targets today. These extensions would no longer be possible.

Reviewer 4: We also need to update KDoc in the corresponding implementation. @dkhalanskyjb

Reviewer 4: > We also need to update KDoc in the corresponding implementation.

Never mind, it looks OK as of now.

`MIN` and `MAX` are only used for comparing to them, and we can check this via
indirect and less efficient means:

```kotlin
fun epochSecondsLaterThanMaxInstant(epochSeconds: Long): Boolean =
   Instant.fromEpochSeconds(epochSeconds) ==
   Instant.fromEpochSeconds(epochSeconds - 1)
```

#### Process 1: pure library solutions

* Standard Library publishes its own `kotlin.time.Instant`.
* `kotlinx-datetime`:
  - And adds conversion functions between `kotlinx.datetime.Instant`
    and `kotlin.time.Instant`, as well as between `kotlinx.datetime.Clock` and
    `kotlin.time.Clock`. These are public.
  - Deprecates `kotlinx.datetime.Instant` and `kotlinx.datetime.Clock`
    with a warning and a replacement with
    `kotlin.time.Instant` and `kotlin.time.Clock`.
  - Deprecates every function that accepts an `Instant`/`Clock`,
    adds an overload that accepts `kotlin.time.(Instant|Clock)`
    instead. The proposed suggestions involve the new converters.
    Exception: `Clock.asTimeSource` does not get the new overload.
  - For every function that returns but doesn't accept an `Instant`,
    hides it and adds a new function, one that returns `kotlin.time.Instant`,

>     but with a different `JvmName`.

Reviewer 2: I think a different `JvmName` isn't needed. Iirc two methods with the same name but different return types can exist in the JVM bytecode.

Author: Yes, you are right. For other platforms, though, it's still needed. Edited.

Instant and Clock.
=================

* **Type**: Standard Library API proposal
* **Author**: Dmitry Khalanskiy
* **Contributors**: Ilya Gorbunov, Roman Elizarov, Vsevolod Tolstopyatov
* **Status**: pending
* **Prototype**: available in `kotlinx-datetime`

Summary
-------

This proposal aims to introduce facilities for obtaining the current system time
into the Kotlin Standard Library.

Additionally, it defines some interactions with `kotlin.time.Duration`, an
existing API for representing some number of seconds.

The API described here is available in `kotlinx-datetime` 0.6.0 in the form of
the `Instant` class and the `Clock` interface.
The goal of this proposal is to move this widely applicable functionality into
the Standard Library.

Overview
--------

### `Instant`

> `Instant` denotes a moment in time, regardless of the observer.

Reviewer 5: Maybe best to remove the phrase "regardless of observer" since that evokes relativistic effects.  According to the theory of relativity, simultaneity is relative.  If events occur in different places, there may be frames of reference in which they are simultaneous, other frames in which one occurs first and other frames in which the other occurred first.

There is no "absolute" frame of reference from which we can say that two events occurring at different places occurred at the "same instant".

So, I'm thinking it's probably best not to mention "observers" since none of our usual language and intuition about instants or clocks works when we try to account for observers in different reference frames.

Author: Doesn't "regardless of the observer" strongly imply that relativistic effects are not accounted for, given that this definition is absurd in the context of relativity?

Reviewer 5: "Regardless of observer" means "for any observer."  Or "in any frame of reference."

But everything else we're talking about here is only true if everyone is in the same frame of reference.

In other words, only for specific observers.

So using the phrase "regardless of observer" is the opposite of what we want.

Probably best just not to mention observers.

> to provide any clock that is sub-second accurate, or that progresses
> monotonically or smoothly.
> Implementations are therefore not required to actually perform the UTC-SLS
> slew or to otherwise be aware of leap seconds.
> JSR-310 does, however, require that implementations must document the approach
> they use when defining a clock representing the current instant.

Conceptually, this means that `Instant` values taken at the exact same physical
moment and obtained from perfectly synchronized but different `Clock`
implementations are allowed to have different representations depending on
whether the `Clock` takes leap seconds into account and performs smearing.

However, practically, the "perfectly synchronized clocks" and "same physical
moment" requirements make this discrepancy impossible to observe unless
someone is actively trying to: this one-second difference between clock readings
occurs much more often due to
[clock drift](https://en.wikipedia.org/wiki/Clock_drift) than it does due to
leap seconds.

<https://github.com/ThreeTen/threeten-extra/tree/4e016340b97cab604114d10e02a672c1e94c6be5/src/main/java/org/threeten/extra/scale>

> is an implementation that converts between "UTC `Instant`", a new class, and

Reviewer 4: @dkhalanskyjb should we use the actual class name here?

> is an implementation that converts between "UTC `Instant`", a new class, and

Author: I don't think the name of the class is important here. If it were called `LeapSecondAwareInstant` or something similar, I'd still prefer calling it a "UTC `Instant`", that is, an `Instant` using the UTC time-scale.

Instant and Clock.
=================

* **Type**: Standard Library API proposal
* **Author**: Dmitry Khalanskiy
* **Contributors**: Ilya Gorbunov, Roman Elizarov, Vsevolod Tolstopyatov
* **Status**: pending

> * **Prototype**: available in `kotlinx-datetime`

Reviewer 6: ```suggestion
* **Prototype**: available in `kotlinx-datetime`
* **Discussion**: [KEEP-395](https://github.com/Kotlin/KEEP/issues/395)
```

`Instant` is a class with straightforward semantics.
Of the use cases listed for `Instant`, almost none require timezone awareness,
so the `Instant` API does not benefit much from being in the same library as
the other `kotlinx.datetime` classes.

**Example**: the Okio library only uses `kotlinx.datetime` for `Instant` and
`Clock`.

**Example**: `Instant` is referenced more often than any of the `Local*`
classes: <https://grep.app/search?q=kotlinx.datetime.Instant> has about 800,
and <https://grep.app/search?q=kotlinx.datetime.Local> about 550 hits as of
writing.

Proposal
--------

The proposal is to remove the `Instant` class and the `Clock` interface
from the `kotlinx-datetime` library and add them to the Standard Library's
`kotlin.time` package.

The specifics of what gets added is given below.

### `Instant` API

```kotlin
/**
 * A moment in time.
 *
 * A point in time must be uniquely identified in a way that is independent of a time zone.
 * For example, `1970-01-01, 00:00:00` does not represent a moment in time since this would happen at different times
 * in different time zones: someone in Tokyo would think it is already `1970-01-01` several hours earlier than someone in

>  * Berlin would. To represent such entities, use the `kotlin-time time.

Reviewer 7: ```suggestion
 * Berlin would. To represent such entities, use `LocalDateTime` from the `kotlinx-datetime` library.
```

         *
         * @see Instant.toString for formatting.
         */
        public fun parse(input: CharSequence): Instant

        /**
         * An instant value that is far in the past.
         *
         * [isDistantPast] returns true for this value and all earlier ones.
         */
        public val DISTANT_PAST: Instant

        /**
         * An instant value that is far in the future.
         *
         * [isDistantFuture] returns true for this value and all later ones.
         */
        public val DISTANT_FUTURE: Instant
    }
}

/**
 * Returns true if the instant is [Instant.DISTANT_PAST] or earlier.
 */
public val Instant.isDistantPast: Boolean
    get() = this <= Instant.DISTANT_PAST

/**
 * Returns true if the instant is [Instant.DISTANT_FUTURE] or later.
 */
public val Instant.isDistantFuture: Boolean
    get() = this >= Instant.DISTANT_FUTURE
```

All parts of this API have existed in `kotlinx-datetime` for a while now and
have stable, widely used and thoroughly tested implementations.

This API is based on the eponymous API entry in JSR 310 (available in

> the 310bp project and the standard library of Java9 and later):

Reviewer 2: Isn't it available since Java 8?

     * - [now] returns `2023-01-02T22:32:05Z`.
     *
     * When you need predictable intervals between successive measurements, consider using [TimeSource.Monotonic].
     *
     * For improved testability, you should avoid using [Clock.System] directly in the implementation
     * and pass a [Clock] explicitly instead.
     */
    public object System : Clock {
        override fun now(): Instant = @Suppress("DEPRECATION_ERROR") Instant.now()
    }

    /** A companion object used purely for namespacing. */
    public companion object
}
```

This class mimics Java's
[Clock](https://docs.oracle.com/javase/8/docs/api/java/time/Clock.html),
but without a time zone attached.
This makes the interface a clock in the sense of "system clock",
without the ability to serve as a wall-clock.

`now()` is not a method on `Instant`, even though that would be more convenient,
in order to encourage writing testable code using dependency injection.
If some function in the depths of the system calls `Instant.now()`, it becomes
nondeterministic and difficult to reliably test, whereas passing a `Clock`
to it makes calling `now()` a functionally pure operation.

> ### Migration process

Reviewer 6: I'd leave a description of that process we've chosen to implement.

Reviewer 6: I meant leave only that description because having others at this point doesn't bring much use.
